be/src/cache/block_cache.cpp

```diff
@@ -0,0 +1,142 @@
+// This file is licensed under the Elastic License 2.0. Copyright 2021-present, StarRocks Limited.
+
+#include "cache/block_cache.h"
+
+#include <fmt/format.h>
+
+#include "common/config.h"
+#include "common/statusor.h"
+#include "common/logging.h"
+#include "gutil/strings/substitute.h"
+#include "cache/fb_cachelib.h"
+
+//#include "util/hash_util.hpp"
+#include "formats/orc/apache-orc/c++/src/Murmur3.hh"
+
+namespace starrocks {
+
+BlockCache::BlockCache() {
+    _kv_cache = new FbCacheLib();
+}
+
+BlockCache::~BlockCache() {
+    if (_kv_cache) {
+        _kv_cache->destroy();
+        delete _kv_cache;
+        _kv_cache = nullptr;
+    }
+}
+
+BlockCache* BlockCache::instance() {
+    static BlockCache cache;
+    return &cache;
+}
+
+Status BlockCache::init(const CacheOptions& options) {
+    return _kv_cache->init(options);
+}
+
+Status BlockCache::set_block_size(size_t block_size) {
+    // TODO: check block size limit
+    _block_size = block_size;
+    return Status::OK();
+}
+
+Status BlockCache::write_cache(const CacheKey& cache_key, off_t offset, size_t size, const char* buffer,
+                               size_t ttl_seconds) {
+    if (offset % _block_size != 0) {
+        LOG(WARNING) << "write block key: " << cache_key << " with invalid args, offset: " << offset;
+    }
+    if (!buffer) {
+        return Status::InvalidArgument("invalid data buffer");
+    }
+    if (size == 0) {
+        return Status::OK();
+    }
+        
+    size_t start_block_index = offset / _block_size;
+    size_t end_block_index = (offset + size - 1) / _block_size + 1;
```

Comment:
offset + _block_size - 1

Reply:
> offset + _block_size - 1

The write_cache operation may cross multiple blocks and we split to single block write here, so end_block_index represents the last block in this write.